### ISSUE TITLE
ubuntu: scylla-tools rcX package causes trouble to upgrade because of versioning scheme

### ISSUE THREAD

**Issue by syuu1228** [scylladb/scylla#2765]
While fixing #2732, I found scylla-tools does not able to upgrade (becomes 'kept back') only when version "2.0~rcX", and it is not occurs when "2.0":

```
The following packages have been kept back:
   scylla-tools (1.7.4-20170726.7a15b00-0ubuntu1 => 2.0~rc3-20170906.389d29b-0ubuntu1~xenial)
The following packages will be upgraded:
   scylla (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
1 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.
Need to get 0 B/13.2 kB of archives.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] 
```

Looks like "~" causes the problem, it can avoid when using "+":
```
The following NEW packages will be installed:
   scylla-tools-core (2.0+rc3-20170906.686a19c-0ubuntu1~xenial)
The following packages will be upgraded:
   scylla (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-tools (1.7.4-20170726.7a15b00-0ubuntu1 => 2.0+rc3-20170906.686a19c-0ubuntu1~xenial)
2 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/23.3 MB of archives.
After this operation, 17.4 kB of additional disk space will be used.
Do you want to continue? [Y/n] 
```

**Comment by avikivity** [comment]
According to [this](https://serverfault.com/questions/604541/debian-packages-version-convention), when comparing `1.7.4-20170726.7a15b00-0ubuntu1` to `2.0~rc3-20170906.389d29b-0ubuntu1~xenial`, the rules are:

 - Compare the epoch: both should be zero
 - Compare `1` to `2` -> `2` wins (no further compares)

So I don't understand why it didn't work.

Changing ~ to + will mean that 2.0 will be before 2.0~rc3, because "~" sorts before "".

**Comment by syuu1228** [comment]
I haven't find any evidence from Debian Policy Manual that our current versioning scheme is wrong...

Maybe our packaging is something wrong, but at least specifying "Replaces: pkg (<< new-ver)" and "Breaks: pkg (<< new-ver)" is a standard way to spiting package into two.

I tried to get debug output on both apt-get and apt, here's a result of debug output.
(debug option is:  -o Debug::pkgDepCache::AutoInstall=true -o Debug::pkgDepCache::Marker=true -o Debug::pkgProblemResolver=true -o Debug::pkgPolicy=true -o  Debug::pkgProblemResolver::ShowScores=true)

To know more details how the apt-get/apt command decided that dependency is not satisfied, I suppose we need to look into apt source code.

### "apt-get -V upgrade" with "~rc3" package
#### normal output
```
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
The following packages have been kept back:
   scylla-server (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-tools (1.7.4-20170726.7a15b00-0ubuntu1 => 2.0~rc3-20170906.1526fec-0ubuntu1~xenial)
The following packages will be upgraded:
   scylla (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-conf (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-jmx (1.7.4-20170726.51ba6ea-0ubuntu1 => 2.0~rc2-20170815.fdfca8e-0ubuntu1~xenial)
   scylla-kernel-conf (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
4 upgraded, 0 newly installed, 0 to remove and 2 not upgraded.
Need to get 5,922 kB of archives.
After this operation, 1,024 B of additional disk space will be used.
Do you want to continue? [Y/n] 
```
#### debug output
```
Entering ResolveByKeep
  Dependencies are not satisfied for scylla-server [ amd64 ] < 1.7.4-0.20170726.ff643e3-0ubuntu1 -> 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial > ( database )
Keeping package scylla-server:amd64
  MarkKeep scylla-server [ amd64 ] < 1.7.4-0.20170726.ff643e3-0ubuntu1 -> 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial > ( database ) FU=0
  Dependencies are not satisfied for scylla-tools [ amd64 ] < 1.7.4-20170726.7a15b00-0ubuntu1 -> 2.0~rc3-20170906.1526fec-0ubuntu1~xenial > ( misc )
Keeping package scylla-tools:amd64
  MarkKeep scylla-tools [ amd64 ] < 1.7.4-20170726.7a15b00-0ubuntu1 -> 2.0~rc3-20170906.1526fec-0ubuntu1~xenial > ( misc ) FU=0
```
### "apt -V upgrade" with "~rc3" package
#### normal output
```
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
The following NEW packages will be installed:
   python3-pyudev (0.16.1-3)
The following packages have been kept back:
   scylla-tools (1.7.4-20170726.7a15b00-0ubuntu1 => 2.0~rc3-20170906.1526fec-0ubuntu1~xenial)
The following packages will be upgraded:
   scylla (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-conf (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-jmx (1.7.4-20170726.51ba6ea-0ubuntu1 => 2.0~rc2-20170815.fdfca8e-0ubuntu1~xenial)
   scylla-kernel-conf (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-server (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
5 upgraded, 1 newly installed, 0 to remove and 1 not upgraded.
Need to get 15.5 MB of archives.
After this operation, 4,903 kB of additional disk space will be used.
Do you want to continue? [Y/n] 
```
#### debug output
```
Entering ResolveByKeep
  Dependencies are not satisfied for scylla-tools-core [ amd64 ] < none -> 2.0~rc3-20170906.1526fec-0ubuntu1~xenial > ( misc )
Keeping package scylla-tools-core:amd64
  MarkKeep scylla-tools-core [ amd64 ] < none -> 2.0~rc3-20170906.1526fec-0ubuntu1~xenial > ( misc ) FU=0
Calculating upgrade... Done
```

### "apt-get -V upgrade" with "+rc3" package (same result without rcX)
#### normal output 
```
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
The following packages have been kept back:
   scylla-server (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-tools (1.7.4-20170726.7a15b00-0ubuntu1 => 2.0+rc3-20170906.9568752-0ubuntu1~xenial)
The following packages will be upgraded:
   scylla (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-conf (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-jmx (1.7.4-20170726.51ba6ea-0ubuntu1 => 2.0~rc2-20170815.fdfca8e-0ubuntu1~xenial)
   scylla-kernel-conf (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
4 upgraded, 0 newly installed, 0 to remove and 2 not upgraded.
Need to get 5,922 kB of archives.
After this operation, 1,024 B of additional disk space will be used.
Do you want to continue? [Y/n] 
```
#### debug output 
```
Entering ResolveByKeep
  Dependencies are not satisfied for scylla-server [ amd64 ] < 1.7.4-0.20170726.ff643e3-0ubuntu1 -> 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial > ( database )
Keeping package scylla-server:amd64
  MarkKeep scylla-server [ amd64 ] < 1.7.4-0.20170726.ff643e3-0ubuntu1 -> 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial > ( database ) FU=0
  Dependencies are not satisfied for scylla-tools [ amd64 ] < 1.7.4-20170726.7a15b00-0ubuntu1 -> 2.0+rc3-20170906.9568752-0ubuntu1~xenial > ( misc )
Keeping package scylla-tools:amd64
  MarkKeep scylla-tools [ amd64 ] < 1.7.4-20170726.7a15b00-0ubuntu1 -> 2.0+rc3-20170906.9568752-0ubuntu1~xenial > ( misc ) FU=0
```
### "apt -V upgrade" with "+rc3" package (same result without rcX)
#### normal output 
```
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
The following NEW packages will be installed:
   python3-pyudev (0.16.1-3)
   scylla-tools-core (2.0+rc3-20170906.9568752-0ubuntu1~xenial)
The following packages will be upgraded:
   scylla (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-conf (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-jmx (1.7.4-20170726.51ba6ea-0ubuntu1 => 2.0~rc2-20170815.fdfca8e-0ubuntu1~xenial)
   scylla-kernel-conf (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-server (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-tools (1.7.4-20170726.7a15b00-0ubuntu1 => 2.0+rc3-20170906.9568752-0ubuntu1~xenial)
6 upgraded, 2 newly installed, 0 to remove and 0 not upgraded.
Need to get 38.8 MB of archives.
After this operation, 4,920 kB of additional disk space will be used.
Do you want to continue? [Y/n] 
```
#### debug output 
```
  MarkInstall scylla-server [ amd64 ] < 1.7.4-0.20170726.ff643e3-0ubuntu1 -> 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial > ( database ) FU=0
  Installing python3-pyudev as Depends of scylla-server
    MarkInstall python3-pyudev [ amd64 ] < none -> 0.16.1-3 > ( python ) FU=0
  MarkInstall scylla-tools [ amd64 ] < 1.7.4-20170726.7a15b00-0ubuntu1 -> 2.0+rc3-20170906.9568752-0ubuntu1~xenial > ( misc ) FU=0
  Installing scylla-tools-core as Depends of scylla-tools
    MarkInstall scylla-tools-core [ amd64 ] < none -> 2.0+rc3-20170906.9568752-0ubuntu1~xenial > ( misc ) FU=0
```

**Comment by syuu1228** [comment]
This is not directly related to "~rcX" problem(does NOT solve the problem) but l think it still important,
"apt upgrade" can also be "apt-get dist-upgrade", too (the behavior looks like same).

According to man page of apt-get, if we want to follow package dependency changes we need to use dist-upgrade instead of upgrade.
https://manpages.debian.org/jessie/apt/apt-get.8.en.html
```
upgrade
Packages currently installed with new versions available are retrieved and upgraded;
under no circumstances are currently installed packages removed, or packages not already installed 
retrieved and installed.
```
```
dist-upgrade
dist-upgrade in addition to performing the function of upgrade, also intelligently handles changing 
dependencies with new versions of packages.
```

I saw many Debian distro user suggest to use "apt-get dist-upgrade" when you get "kept back", probably it's common way.

**Comment by syuu1228** [comment]
@avikivity @penberg I think I found the real reason why the issue happens.

We have "dpkg --compare-versions" to compare two package version:
https://manpages.debian.org/jessie/dpkg/dpkg.1.en.html
https://debian-handbook.info/browse/stable/sect.manipulating-packages-with-dpkg.html

It can be use as follows:
```
$ dpkg --compare-versions 1.2-3 gt 1.1-4
$ echo $?
0
$ dpkg --compare-versions 1.2-3 lt 1.1-4
$ echo $?
1
```

Now we have following definition on our debian/control, so our 2.0 package should be "greater than 2.0".
```
Replaces: scylla-tools (<< 2.0)
Breaks: scylla-tools (<< 2.0)
```

However, By testing dpkg --compare-versions, I got following:
```
$ dpkg --compare-versions 2.0~rc3-20170906.389d29b-0ubuntu1~xenial gt 2.0 ; echo $?
1
$ dpkg --compare-versions 2.0-20170906.389d29b-0ubuntu1~xenial gt 2.0 ; echo $?
0
$ dpkg --compare-versions 2.0~rc3-20170906.389d29b-0ubuntu1~xenial lt 2.0 ; echo $?
0
```

so 2.0~rc3 is less than 2.0.
It looks like we should include ~rc versions, then.
I've found following pattern:
```
$ dpkg --compare-versions 2.0~rc3-20170906.389d29b-0ubuntu1~xenial gt 2.0~ ; echo $?
0
$ dpkg --compare-versions 2.0-20170906.389d29b-0ubuntu1~xenial gt 2.0~ ; echo $?
0
$ dpkg --compare-versions 2.0~rc3-20170906.389d29b-0ubuntu1~xenial lt 2.0~ ; echo $?
1
```

I guess it means we should rewrite Replaces/Breaks pattern like following:
```
Replaces: scylla-tools (<< 2.0~)
Breaks: scylla-tools (<< 2.0~)
```

**Comment by syuu1228** [comment]
Tested with "Replaces/Breaks: scylla-tools (<< 2.0~)", but doesn't work:

```
$ sudo apt-get -V dist-upgrade
Reading package lists... Done
Building dependency tree
Reading state information... Done
Calculating upgrade... Done
The following package was automatically installed and is no longer required:
   collectd (5.5.1-1build2)
Use 'sudo apt autoremove' to remove it.
The following packages will be upgraded:
   scylla (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-conf (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-jmx (1.7.4-20170726.51ba6ea-0ubuntu1 => 2.0~rc2-20170815.fdfca8e-0ubuntu1~xenial)
   scylla-kernel-conf (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
   scylla-server (1.7.4-0.20170726.ff643e3-0ubuntu1 => 2.0~rc2-0.20170815.8370e1b-0ubuntu1~xenial)
5 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 15.5 MB of archives.
After this operation, 4,724 kB of additional disk space will be used.
Do you want to continue? [Y/n]
```
Debug output:
```
  scylla-tools:amd64 Depends on scylla-tools-core [ amd64 ] < none -> 2.0~rc3-20170907.4ba7def-0ubuntu1~xenial > ( misc ) (>= 2.0) can't be satisfied!
```

**Comment by slivne** [comment]
fixed by scylladb/scylla-tools-java@f786892372001ffa769d7a6dcddeb50c774899ae
